File: 0279-perfect-squares/0279-perfect-squares.kt
class Solution {
    fun numSquares(n: Int): Int {
        val ar = IntArray(n+1) { i -> i };

        for(i in 4..n){
            var j = 0;
            while(j*j <= i){
                ar[i] = Math.min(ar[i], ar[i-(j*j)]+1);
                // println(ar.joinToString())
                j++;
            }
        }
        return ar[n];
    }
}

File: 0279-perfect-squares/0279-perfect-squares.js
/**
 * @param {number} n
 * @return {number}
 */
var numSquares = function(n) {
    var ar = [];
    for(var i = 0; i <= n; i++) ar.push(i);

    for(var i = 0; i <=n; i++){
        for (var j = 0; j*j <= i; j++){
            ar[i] = Math.min(ar[i], ar[i - (j*j)]+1);
            // console.log(ar);
        }
    }
    return ar[n];
};